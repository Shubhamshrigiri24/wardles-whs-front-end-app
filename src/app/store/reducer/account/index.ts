import { handleActions } from "redux-actions";
import { AnyAction, combineReducers } from "redux";

import { ACTION_STATUSES, TActionStatus } from "app/store/types";

import { State as AccountState, TAccount, AccountActionTypes } from "./types";
import { ACCOUNT_ACTIONS } from "app/store/reducer/account/constants";

const defaultAccountState: AccountState = {
  data: null,
  status: ACTION_STATUSES.INITIAL,
};

const customerActions = handleActions(
  {
    [ACCOUNT_ACTIONS.SET_ACCOUNT]: (
      _state: TAccount | null,
      action: AccountActionTypes
    ) => {
      const a = action as AnyAction;
      const { payload } = a;

      return payload;
    },
  },
  defaultAccountState.data
);

const statusActions = handleActions(
  {
    [ACCOUNT_ACTIONS.SET_ACCOUNT_STATUS]: (
      _state: TActionStatus,
      action: AccountActionTypes
    ) => {
      const a = action as AnyAction;
      const { payload } = a;

      return payload;
    },
  },
  defaultAccountState.status
);

const reducer = combineReducers({
  data: customerActions,
  status: statusActions,
});

export default reducer;
