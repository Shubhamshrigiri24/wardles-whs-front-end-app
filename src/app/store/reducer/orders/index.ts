import { handleActions } from "redux-actions";
import { AnyAction, Reducer, combineReducers } from "redux";

import { ACTION_STATUSES, TActionStatus } from "app/store/types";
import { State as OrdersState, TOrders } from "./types";
import { ORDERS_ACTIONS } from "app/store/reducer/orders/constants";

const defaultOrdersState: OrdersState = {
  data: [],
  status: ACTION_STATUSES.INITIAL,
};

const ordersDataActions = handleActions(
  {
    [ORDERS_ACTIONS.SET_ORDERS]: (_state: TOrders, action: AnyAction) => {
      return action.payload;
    },
    [ORDERS_ACTIONS.SET_ORDER]: (_state: TOrders | null, action: AnyAction) => {
      const a = action as AnyAction;
      const { payload } = a;
      const newOrdersArray: TOrders = [...(_state || [])];

      newOrdersArray.push(payload.order);

      return newOrdersArray;
    },
  },
  defaultOrdersState.data
);

const statusActions = handleActions(
  {
    [ORDERS_ACTIONS.SET_ORDERS_STATUS]: (
      _state: TActionStatus,
      action: AnyAction
    ) => {
      return action.payload;
    },
  },
  defaultOrdersState.status
);

const reducer: Reducer<OrdersState, any> = combineReducers({
  data: ordersDataActions,
  status: statusActions,
});

export default reducer;
