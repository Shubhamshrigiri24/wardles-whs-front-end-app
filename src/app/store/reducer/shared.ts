import { RSAA } from "redux-api-middleware";

export interface APICallAction {
  type: string;
  [RSAA]: {
    endpoint: string;
    method: string;
    types: [string, string, string];
    body?: string;
  };
}

export interface PostcodeLookupAddress {
  line1: string;
  line2?: string;
  city: string;
  postcode: string;
  country: string;
}

export const defaultPostcodeLookup: PostcodeLookupAddress = {
  line1: "",
  city: "",
  postcode: "",
  country: "",
};

export const queryStringDateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'";
