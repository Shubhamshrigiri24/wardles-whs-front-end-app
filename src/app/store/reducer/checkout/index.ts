import { handleActions } from "redux-actions";
import { AnyAction, combineReducers } from "redux";
import {
  CustomerDetails,
  AddCustomerDetailsAction,
  FetchBraintreeTokenResponseAction,
  FetchAddressFromPostcodeActionResponse,
  DeliveryDetails,
  AddDeliveryDetailsAction,
} from "app/store/reducer/checkout/types";
import {
  ADD_CUSTOMER_DETAILS,
  ADD_DELIVERY_DETAILS,
  FETCH_ADDRESS_FROM_POSTCODE_FAILURE,
  FETCH_ADDRESS_FROM_POSTCODE_RESPONSE,
  FETCH_BRAINTREE_TOKEN_REQUEST,
  FETCH_BRAINTREE_TOKEN_RESPONSE,
  emptyDeliveryDetails,
} from "app/store/reducer/checkout/constants";

import { PostcodeLookupAddress } from "../shared";

const defaultCustomerDetails: CustomerDetails = {
  firstName: "",
  lastName: "",
  email: "",
  postcode: "",
  line1: "",
  line2: "",
  city: "",
  phone: "",
  surgeryName: "",
  surgeryAddress: "",
  allowMarketing: false,
};

const customerDetails = handleActions(
  {
    [ADD_CUSTOMER_DETAILS]: (_state: CustomerDetails, action: AnyAction) => {
      const a = action as AddCustomerDetailsAction;
      return a.payload.customerDetails;
    },
  },
  defaultCustomerDetails
);

const deliveryDetails = handleActions(
  {
    [ADD_DELIVERY_DETAILS]: (
      _state: DeliveryDetails,
      action: AddDeliveryDetailsAction
    ) => {
      return action.payload.deliveryDetails;
    },
  },
  emptyDeliveryDetails
);

const braintreeToken = handleActions(
  {
    [ADD_CUSTOMER_DETAILS]: () => "",
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [FETCH_BRAINTREE_TOKEN_RESPONSE]: (_state: string, action: AnyAction) => {
      const a = action as FetchBraintreeTokenResponseAction;
      return a.payload.token;
    },
  },
  ""
);

const paymentErrorMessage = handleActions(
  {
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
  },
  ""
);

const confirmedPayment = handleActions(
  {
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => false,
    [ADD_CUSTOMER_DETAILS]: () => false,
  },
  false
);

const postCodeLookUpAddresses = handleActions(
  {
    [FETCH_ADDRESS_FROM_POSTCODE_RESPONSE]: (
      _state: PostcodeLookupAddress[],
      action: AnyAction
    ) => {
      const a = action as FetchAddressFromPostcodeActionResponse;
      return a.payload.addresses;
    },
    [FETCH_ADDRESS_FROM_POSTCODE_FAILURE]: () => [],
  },
  []
);

const reducer = combineReducers({
  customerDetails,
  confirmedPayment,
  braintreeToken,
  paymentErrorMessage,
  postCodeLookUpAddresses,
  deliveryDetails,
});

export default reducer;
