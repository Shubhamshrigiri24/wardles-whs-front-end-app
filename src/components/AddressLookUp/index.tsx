import React, { useCallback, useState } from "react";
import {
  TextField,
  makeStyles,
  Button,
  Grid,
  Popper,
  List,
  ListItemText,
  ListItem,
  Paper,
  ClickAwayListener,
  useTheme,
  useMediaQuery,
} from "@material-ui/core";
import { PostcodeLookupAddress } from "../../app/store/reducer/shared";

const useStyles = makeStyles((theme) => ({
  input: {
    width: "100%",
  },
  searchButton: {
    width: "100%",
    height: "100%",
    maxHeight: "56px",
  },
  popper: {
    border: "1px solid #AFB7BC",
    backgroundColor: theme.palette.background.paper,
    zIndex: 99,
    borderRadius: "1px 1px 8px 8px",
    maxHeight: 350,
    overflowY: "auto",
  },
}));

export interface AddressLookUpProps {
  fetchAddressesByPostcode(p: string): void;

  postCodeLookUpAddresses: PostcodeLookupAddress[];

  setAddress(postcodeLookupAddress: PostcodeLookupAddress): void;

  className?: string;
  error: string;
  inputRef?: any;
}

const AddressLookUp: React.FC<AddressLookUpProps> = (props) => {
  const {
    fetchAddressesByPostcode,
    postCodeLookUpAddresses,
    setAddress,
    className,
    error,
    inputRef,
  } = props;
  // eslint-disable-next-line @typescript-eslint/naming-convention
  const [open, setOpen] = useState<boolean>(false);
  const [postcode, setPostcode] = useState("");
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down("sm"));

  const handlePostCodeChange = useCallback(
    (event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) => {
      const pc = event.target.value;
      setPostcode(pc);
    },
    [setPostcode]
  );

  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);

  const [popperWidth, setPopperWidth] = useState<undefined | number>(0);

  const onSearchClick = useCallback(() => {
    setAnchorEl(document.getElementById("postcode-input"));
    setPopperWidth(document.getElementById("postcode-input")?.scrollWidth);
    setOpen(true);
    fetchAddressesByPostcode(postcode);
  }, [setOpen, fetchAddressesByPostcode, postcode]);

  const onPostcodeSelectClick = useCallback(
    (postCodeAddress: PostcodeLookupAddress) => () => {
      setAddress(postCodeAddress);
      setOpen(false);
    },
    [setOpen, setAddress]
  );

  const handleClickAway = useCallback(() => {
    setOpen(false);
  }, []);

  const classes = useStyles();
  return (
    <Grid container spacing={2} className={className}>
      <Grid item xs={isMobile ? 12 : 9}>
        <TextField
          className={classes.input}
          label={"Postcode"}
          variant={"outlined"}
          name={"postcode"}
          onChange={handlePostCodeChange}
          InputLabelProps={{
            shrink: true,
          }}
          inputProps={{
            "data-testid": "postcode",
          }}
          id={"postcode-input"}
          error={!!error}
          helperText={error}
          inputRef={inputRef}
        />
        {open && (
          <ClickAwayListener onClickAway={handleClickAway}>
            <Popper
              open={open && postCodeLookUpAddresses.length > 0}
              anchorEl={anchorEl}
              placement={"bottom"}
              className={classes.popper}
              style={{ width: popperWidth }}
            >
              <Paper elevation={10}>
                <List component={"nav"}>
                  {postCodeLookUpAddresses.map((address) => {
                    return (
                      <ListItem
                        button
                        onClick={onPostcodeSelectClick(address)}
                        key={`${address.postcode}+${address.line1}`}
                      >
                        <ListItemText
                          primary={`${address.line1}, ${address.line2}${
                            address.line2 ? ", " : ""
                          }${address.city}, ${address.postcode}`}
                        />
                      </ListItem>
                    );
                  })}
                </List>
              </Paper>
            </Popper>
          </ClickAwayListener>
        )}
      </Grid>
      <Grid item xs={isMobile ? 6 : 3}>
        <Button
          data-testid={"addressLookUp/search-button"}
          variant={"outlined"}
          color={"default"}
          className={classes.searchButton}
          onClick={onSearchClick}
          disabled={postcode.length < 3 || !!error}
        >
          Search
        </Button>
      </Grid>
    </Grid>
  );
};

export default AddressLookUp;
