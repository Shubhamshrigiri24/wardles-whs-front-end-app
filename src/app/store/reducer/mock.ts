import faker from "faker/locale/en_GB";
import { CustomerDetails, DeliveryDetails } from "./checkout/types";
import { PostcodeLookupAddress } from "./shared";
import {
  OnlineQuestion,
  Pack,
  Product,
  SuccessfulConsultationResponse,
} from "./online/types";

export function makeCustomerDetails(
  override?: Partial<CustomerDetails>
): CustomerDetails {
  return {
    firstName: "firstName",
    lastName: "lastName",
    email: "a@b.com",
    postcode: "M19 2UL",
    line1: "line1",
    line2: "line2",
    city: "city",
    phone: "01189998819991197253",
    surgeryName: "surgeryName",
    surgeryAddress: "surgeryAddress",
    allowMarketing: true,
    ...override,
  };
}

export function makeDeliveryDetails(): DeliveryDetails {
  return {
    title: "Mrs",
    firstName: faker.name.firstName(),
    lastName: faker.name.lastName(),
    postcode: faker.address.zipCode(),
    line1: faker.address.streetName(),
    line2: faker.address.streetName(),
    city: faker.address.city(),
  };
}

export function makePostCodeLookUpAddresses(): PostcodeLookupAddress[] {
  return [
    {
      line1: "64 Broom Avenue",
      line2: "",
      city: "Manchester",
      country: "England",
      postcode: "M19 2UL",
    },
    {
      line1: "66 Broom Avenue",
      line2: "",
      city: "Manchester",
      country: "England",
      postcode: "M19 2UL",
    },
    {
      line1: "68 Broom Avenue",
      line2: "",
      city: "Manchester",
      country: "England",
      postcode: "M19 2UL",
    },
  ];
}

export function makeSuccessfulConsultationResponse(
  id: string = "default",
  override?: Partial<SuccessfulConsultationResponse>
): SuccessfulConsultationResponse {
  return {
    answer: `answer-${id}`,
    answerString: `answerString-${id}`,
    initialQuestion: `question-${id}`,
    kind: `trepetto-input-type-${id}`,
    question: `question-${id}`,
    type: "string",
    ...override,
  };
}

export function makeOnlineQuestion(number: number = 1): OnlineQuestion {
  return {
    number,
    question: `question-${number}`,
    answer: `answer-${number}`,
    type: "string",
  };
}

export function makeOnlineQuestions(): OnlineQuestion[] {
  return [makeOnlineQuestion(1), makeOnlineQuestion(2), makeOnlineQuestion(3)];
}

export function makeProduct(
  id: string = "default",
  override?: {
    id?: string;
    name?: string;
    service?: string;
    packs?: Pack[];
    variant?: string;
  }
): Product {
  return {
    id,
    name: `name-${id}`,
    service: `service-${id}`,
    packs: [
      makePack(`${id}-pack-1`),
      makePack(`${id}-pack-2`),
      makePack(`${id}-pack-3`),
    ],
    variant: "",
    ...override,
  };
}

export function makePack(
  id: string = "default",
  override?: {
    sku?: string;
    name?: string;
    label?: string;
    price?: number;
    pricePerUnit?: number;
    itemsPerPack?: number;
  }
): Pack {
  return {
    sku: `sku-${id}`,
    name: `name-${id}`,
    label: `label-${id}`,
    basePrice: 1,
    price: 1,
    pricePerUnit: 1,
    itemsPerPack: 1,
    discounts: [],
    ...override,
  };
}

export const customerDetails: CustomerDetails = {
  firstName: "FIRST_NAME",
  lastName: "LAST_NAME",
  email: "EMAIL",
  postcode: "POSTCODE",
  line1: "LINE_1",
  line2: "LINE_2",
  city: "CITY",
  phone: "PHONE",
  surgeryName: "SURGERY_NAME",
  surgeryAddress: "SURGERY_ADDRESS",
  allowMarketing: true,
};

export const deliveryDetails: DeliveryDetails = {
  title: "Mrs",
  firstName: "FIRST_NAME",
  lastName: "LAST_NAME",
  postcode: "POSTCODE",
  line1: "LINE_1",
  line2: "LINE_2",
  city: "CITY",
};
