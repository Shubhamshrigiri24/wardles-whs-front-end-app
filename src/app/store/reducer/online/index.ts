import { handleActions } from "redux-actions";
import {
  allOnlineServices,
  Ed,
  State,
  Pack,
  Product,
  defaultProduct,
  defaultPack,
  SET_PACK,
  SET_PRODUCT,
  SetPackAction,
  SetProductAction,
  FetchProductsResponseAction,
  FETCH_PRODUCTS_RESPONSE,
  FETCH_PRODUCTS_FAILURE,
  FETCH_PRODUCTS_REQUEST,
  FETCH_BRAINTREE_TOKEN_REQUEST,
  FETCH_BRAINTREE_TOKEN_RESPONSE,
  ON_ORDER_FAILURE,
  ON_ORDER_REQUEST,
  ON_ORDER_RESPONSE,
  ON_ORDER,
  CLEAR_ORDER_FAILURE_ERROR,
  OrderResponseAction,
  SetConsultationAction,
  SetIsConsultationValidForOrderAction,
  OnlineQuestion,
  SetOnlineServiceAction,
  OnlineService,
  SET_ONLINE_SERVICE,
  CHECK_PRODUCT_TOKEN_RESPONSE,
  CHECK_PRODUCT_TOKEN_FAILURE,
  CHECK_PRODUCT_TOKEN_REQUEST,
  CheckProductTokenResponseAction,
  SET_PRODUCT_TOKEN,
  SetProductTokenAction,
  INVALIDATE_PRODUCT_TOKEN,
  SEND_CONSULTATION_REQUEST,
  SEND_CONSULTATION_RESPONSE,
  SEND_CONSULTATION_FAILURE,
  SendConsultationSuccessfulResponse,
} from "./types";
import { combineReducers, Reducer } from "redux";
import { GENERIC_ERROR_MESSAGE } from "app/store/config";
import {
  FetchBraintreeTokenResponseAction,
  PaymentFailureAction,
  SET_CONSULTATION,
  SET_IS_CONSULTATION_VALID_FOR_ORDER,
  START_TRIPETTO,
} from "./types";
import { ORDER_ACTIONS } from "app/store/reducer/order/constants";

const selectedOnlineService = handleActions(
  {
    [SET_ONLINE_SERVICE]: (
      _state: OnlineService,
      action: SetOnlineServiceAction
    ) => {
      return action.payload ? action.payload.selectedOnlineService : Ed;
    },
  },
  Ed
);

const onlineServices = handleActions({}, allOnlineServices);

const selectedPack = handleActions(
  {
    [SET_PACK]: (_state: Pack, action: SetPackAction) => {
      return action.payload ? action.payload.selectedPack : defaultPack;
    },
  },
  defaultPack
);

const selectedProduct = handleActions(
  {
    [SET_PRODUCT]: (_state: Product, action: SetProductAction) => {
      return action.payload ? action.payload.selectedProduct : defaultProduct;
    },
  },
  defaultProduct
);

const products = handleActions(
  {
    [FETCH_PRODUCTS_RESPONSE]: (
      _state: Product[],
      action: FetchProductsResponseAction
    ) => {
      return action.payload ? action.payload : [];
    },
    [FETCH_PRODUCTS_FAILURE]: () => [],
    [FETCH_PRODUCTS_REQUEST]: () => [],
    [START_TRIPETTO]: () => [],
  },
  [defaultProduct]
);

const productToken = handleActions(
  {
    [SET_PRODUCT_TOKEN]: (
      _state: string | null,
      action: SetProductTokenAction
    ) => action.payload,
    [INVALIDATE_PRODUCT_TOKEN]: (_state: string | null) => null,
  },
  null
);

const productTokenValid = handleActions(
  {
    [CHECK_PRODUCT_TOKEN_RESPONSE]: (
      _state: boolean,
      action: CheckProductTokenResponseAction
    ) => {
      return !!action.payload.status;
    },
    [CHECK_PRODUCT_TOKEN_FAILURE]: () => false,
    [CHECK_PRODUCT_TOKEN_REQUEST]: () => true,
  },
  true
);

const fetching = handleActions(
  {
    [FETCH_PRODUCTS_REQUEST]: () => true,
    [FETCH_PRODUCTS_RESPONSE]: () => false,
    [FETCH_PRODUCTS_FAILURE]: () => false,
  },
  true
);

const braintreeToken = handleActions(
  {
    [ON_ORDER]: () => "",
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [FETCH_BRAINTREE_TOKEN_RESPONSE]: (
      _state: string,
      action: FetchBraintreeTokenResponseAction
    ) => action.payload.token,
    [START_TRIPETTO]: () => "",
  },
  ""
);

const orderErrorMessage = handleActions(
  {
    [ON_ORDER_FAILURE]: (_state: string, action: PaymentFailureAction) =>
      action?.payload?.response?.message || GENERIC_ERROR_MESSAGE,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [ON_ORDER]: () => "",
    [CLEAR_ORDER_FAILURE_ERROR]: () => "",
  },
  ""
);

const successfulOrder = handleActions(
  {
    [ON_ORDER_FAILURE]: () => false,
    [ON_ORDER_REQUEST]: () => false,
    [SET_CONSULTATION]: () => false,
    [ON_ORDER_RESPONSE]: () => true,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => false,
  },
  false
);

const consultationResponses = handleActions(
  {
    [SET_CONSULTATION]: (
      _state: OnlineQuestion[],
      action: SetConsultationAction
    ) => action?.payload?.consultation || [],
    [START_TRIPETTO]: () => [],
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => [],
  },
  []
);

const consultationToken = handleActions(
  {
    [SEND_CONSULTATION_REQUEST]: () => null,
    [SEND_CONSULTATION_RESPONSE]: (
      _,
      action: SendConsultationSuccessfulResponse
    ) => action.payload ?? null,
    [SEND_CONSULTATION_FAILURE]: () => null,
    [START_TRIPETTO]: () => null,
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => null,
  },
  null
);

const isConsultationValidForOrder = handleActions(
  {
    [SET_IS_CONSULTATION_VALID_FOR_ORDER]: (
      state: boolean,
      action: SetIsConsultationValidForOrderAction
    ) => action.payload,
    [START_TRIPETTO]: () => false,
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => false,
  },
  false
);

const consultation = combineReducers({
  responses: consultationResponses,
  token: consultationToken,
  isValidForOrder: isConsultationValidForOrder,
});

const orderNumber = handleActions(
  {
    [ON_ORDER_RESPONSE]: (_state: string, action: OrderResponseAction) =>
      action?.payload?.orderNumber,
    [FETCH_BRAINTREE_TOKEN_REQUEST]: () => "",
    [ON_ORDER]: () => "",
  },
  ""
);

const reducer: Reducer<State, any> = combineReducers({
  selectedOnlineService,
  onlineServices,
  selectedPack,
  selectedProduct,
  products,
  productToken,
  productTokenValid,
  fetching,
  braintreeToken,
  successfulOrder,
  orderErrorMessage,
  consultation,
  orderNumber,
});

export default reducer;
