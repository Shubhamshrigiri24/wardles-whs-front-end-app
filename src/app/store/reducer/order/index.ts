import { handleActions } from "redux-actions";
import { AnyAction, Reducer, combineReducers } from "redux";

import {
  State as OrderState,
  OrderCustomer,
  OrderProduct,
  SetCustomerActionType,
  SetProductActionType,
} from "app/store/reducer/order/types";
import { START_TRIPETTO } from "app/store/reducer/online/types";
import {
  ORDER_ACTIONS,
  CUSTOMER_TYPES,
} from "app/store/reducer/order/constants";

const defaultOrderState: OrderState = {
  customer: { customerType: CUSTOMER_TYPES.GUEST, customerDetails: null },
  product: null,
};

const customerActions = handleActions(
  {
    [ORDER_ACTIONS.SET_CUSTOMER]: (
      _state: OrderCustomer,
      action: AnyAction
    ) => {
      const a = action as SetCustomerActionType;
      const { payload } = a;

      return payload;
    },
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => defaultOrderState.customer,
    [START_TRIPETTO]: () => defaultOrderState.customer,
  },
  defaultOrderState.customer
);

const productActions = handleActions(
  {
    [ORDER_ACTIONS.SET_PRODUCT]: (_state: OrderProduct, action: AnyAction) => {
      const a = action as SetProductActionType;
      const { payload } = a;

      return payload;
    },
    [ORDER_ACTIONS.CLEANUP_ORDER_FLOW_STATE]: () => defaultOrderState.product,
    [START_TRIPETTO]: () => defaultOrderState.product,
  },
  defaultOrderState.product
);

const reducer: Reducer<OrderState, any> = combineReducers({
  customer: customerActions,
  product: productActions,
});

export default reducer;
