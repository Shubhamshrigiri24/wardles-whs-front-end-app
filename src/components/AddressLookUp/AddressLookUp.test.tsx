import React from "react";
import { render, fireEvent, act } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import AddressLookUp, { AddressLookUpProps } from "./";
import { makePostCodeLookUpAddresses } from "app/store/reducer/mock";

global.document.createRange = () =>
  ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: "BODY",
      ownerDocument: document,
    },
  } as any);

describe("<AddressLookUp />", () => {
  it("should match the snapshot", () => {
    const props = makeProps();
    const { container } = render(<AddressLookUp {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("should search for the postcode when the search button is clicked", async () => {
    const props = makeProps();
    props.error = "";
    const { getByTestId } = render(<AddressLookUp {...props} />);
    const postcode = "M19 2UL";
    await act(async () => {
      fireEvent.change(getByTestId("postcode"), {
        target: { value: postcode },
      });
    });

    await act(async () => {
      fireEvent.click(getByTestId("addressLookUp/search-button"));
    });

    expect(props.fetchAddressesByPostcode).toHaveBeenCalledWith(postcode);
  });
});

function makeProps(): AddressLookUpProps {
  return {
    fetchAddressesByPostcode: jest.fn(),
    postCodeLookUpAddresses: makePostCodeLookUpAddresses(),
    setAddress: jest.fn(),
    className: "className",
    error: "test error message",
  };
}
