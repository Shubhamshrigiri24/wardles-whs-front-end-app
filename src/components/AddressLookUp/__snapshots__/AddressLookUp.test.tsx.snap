// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddressLookUp /> should match the snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root className MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-input-1"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-error Mui-error"
          data-shrink="true"
          for="postcode-input"
          id="postcode-input-label"
        >
          Postcode
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root Mui-error Mui-error MuiInputBase-formControl"
        >
          <input
            aria-describedby="postcode-input-helper-text"
            aria-invalid="true"
            class="MuiInputBase-input MuiOutlinedInput-input"
            data-testid="postcode"
            id="postcode-input"
            name="postcode"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-4 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-6 PrivateNotchedOutline-legendNotched-7"
            >
              <span>
                Postcode
              </span>
            </legend>
          </fieldset>
        </div>
        <p
          class="MuiFormHelperText-root MuiFormHelperText-contained Mui-error"
          id="postcode-input-helper-text"
        >
          test error message
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-searchButton-2 Mui-disabled Mui-disabled"
        data-testid="addressLookUp/search-button"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Search
        </span>
      </button>
    </div>
  </div>
</div>
`;
